src/instrumenter/annotations.ts

```diff
@@ -440,10 +439,9 @@ class AnnotationExtractor {
         const result: AnnotationMetaData[] = [];
 
         const rx =
-            /\s*(\*|\/\/\/)\s*#?(if_succeeds|if_updated|if_assigned|invariant|assert|define\s*[a-zA-Z0-9_]*\s*\([^)]*\))/g;
+            /\s*(\*|\/\/\/)\s*((.*)@custom:scribble\s*)?#?(if_succeeds|if_updated|if_assigned|invariant|assert|define\s*[a-zA-Z0-9_]*\s*\([^)]*\))/g;
```

Comment:
I think the goal of the custom tag would be to utilise natspec to provide it, as opposed to using regular expressions.

However if you keep using regular expressions, it shouldn't allow leading non-whitespace for the custom tag.

Reply:
Thanks for reviewing @axic !

I haven't looked at the PR yet - will let you know early next week (sorry working on another project these past 2-3 weeks). From a cursory glance, I think our current implementation just extends our internal regex-y logic to also parse out `@custom` natspec tags instead of relying on the compiler, which I agree is counter-productive to some degree. 

I think the reason we went with this approach for now, was that in order to provide accurate source line information for various errors in the annotation (syntax errors, type errors, etc..) we need to know the relative offset of the start of the annotation to the beginning of the file. If we do it with regex-es ourselves we know exactly where it is in the file.

As far as I could tell, in the parsed natspec information provided in the compiler we couldn't find any indication what was the offset of the annotation relative to the begining of the file (or even the comment). Do you have any ideas how to overcome this?

Reply:
> I think the reason we went with this approach for now, was that in order to provide accurate source line information for various errors in the annotation (syntax errors, type errors, etc..) we need to know the relative offset of the start of the annotation to the beginning of the file. If we do it with regex-es ourselves we know exactly where it is in the file.
>
> As far as I could tell, in the parsed natspec information provided in the compiler we couldn't find any indication what was the offset of the annotation relative to the begining of the file (or even the comment). Do you have any ideas how to overcome this?

Hmm, this is a good question.  I think the AST view should give you proper source location next to the natspec tags.  (Not only the natspec output, but the AST also contains it under the `documentation` tag.)

Here's an example: https://github.com/ethereum/solidity/blob/develop/test/libsolidity/ASTJSON/documentation.sol and https://github.com/ethereum/solidity/blob/develop/test/libsolidity/ASTJSON/documentation.json

> From a cursory glance, I think our current implementation just extends our internal regex-y logic to also parse out @custom natspec tags instead of relying on the compiler, which I agree is counter-productive to some degree.

If implementing support for the AST would take a longer time, I'd still suggest to strengthen the regex to match the behaviour of upstream, that is not allowing non-whitespace in front of the tag.

Reply:
Why do you need the location relative in the documentation?

Can't you have two code paths, one for natspec based, and the second for your regex based one?

Reply:
We could have two paths. I think the problem is it would be bad user experience if we provide accurate source-line information for errors in old-style annotations, but not for natspec annotations.

Reply:
It should be accurate too with the AST. It tells you the exact position (column/line, but you need to translate it).

Can you please explain what exactly is the issue?

Reply:
Say we have this code:

```
/// blah blah blah
/// blah
/// @custom:scribble if_succeeds x + y > 0;
function foo(uint x, int y) {
...
}
```

In the ast we know that the whole docstring start at say line 5, col 3. This means that the first 'blah' is around line 5 col 3. However we don't know how much further down the `@custom:scribble` annotation starts. When we parse the contents of the annotation itself, and build an AST we know that the "x+y" starts around 29 characters from the "@". 

Now there is a type error at the "x+y". We know that this is 29 charcters relative to the "@", and we know that the "@" is somewhere inside the docstring that starts at line 5 col 3, but we don't know *exactly* where inside that doc string. For this reason we can't easily convert the error offset "29 characters from the @" into the global error offset "X characters from the start of the file"

Reply:
I guess we can do an additional pass over each doc string, looking up the starting index of each of the natspec strings inside. Since we know the full text of each natspec string we should be able to disambiguate. The only edge case is if a user enters _exactly_ the same natspec string twice, however we can just make that an error.

Reply:
We also support legacy Solidity versions, that do not have `@cusom:whatever`. While we are required to do this, we are unable to change approach. Also we have logic that rebuild `StructuredDocumentation` from source, for 0.4 version that do not have this node in AST. This is an additional historical point to highlight,

Reply:
Btw just checked upstream and I think it may be bug that it allows `blabla @cusom:tag test`, because it ignores `blabla` entirely. See https://github.com/ethereum/solidity/issues/12245.

This may be also the reason the offsets may be wrong.

Reply:
For now we are using our old regex/parser approach, as:

1) We need to use it anyways to keep backwards compatibility with older scribble behavior
2) It gives us precise error locations relative to the start of the file (supporting this with natspec will require adding more code)

We changed the behavior to disallow non-white space before the `@custom:scribble`